### PR TITLE
Add 2PC module and coordinator implementation

### PR DESCRIPTION
**Pull request by vaporos**
There are two pieces to this algorithm -- coordinator and participant; this covers the coordinator half.

### CODE REVIEW

File: libaugrim/src/algorithm/two_phase_commit/coordinator_algorithm.rs [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+                    self.push_advance_epoch_actions(context, &mut actions);
+                } else {
+                    self.push_abort_actions(context, &mut actions);
+                }
+
+                Ok(actions)
+            }
+
+            // An alarm may be sent if we've previously used the `CoordinatorAction::Update` action
+            // to set an alsrm to `CoordinatorActionAlarm::Set`, as is the case when we enter the
```
Comment (agunde406) [inline libaugrim/src/algorithm/two_phase_commit/coordinator_algorithm.rs]: alsrm -> alarm

File: libaugrim/src/algorithm/two_phase_commit/coordinator_context.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    T: Time,
+{
+    WaitingForStart,
+    Voting { vote_timeout_start: T },
+    WaitingForVote,
+    Abort,
+    Commit,
+}
+
+// does this need to keep V? (probably)
```
Comment (agunde406) [inline libaugrim/src/algorithm/two_phase_commit/coordinator_context.rs]: This comment can be removed

**Review by agunde406 — CHANGES_REQUESTED** [review]

File: libaugrim/src/algorithm/two_phase_commit/coordinator_algorithm.rs (hunk 1 above)
Comment (vaporos) [reply libaugrim/src/algorithm/two_phase_commit/coordinator_algorithm.rs]: fixed

File: libaugrim/src/algorithm/two_phase_commit/coordinator_context.rs (hunk 2 above)
Comment (vaporos) [reply libaugrim/src/algorithm/two_phase_commit/coordinator_context.rs]: fixed

**Review by agunde406 — APPROVED** [review]

**Review by peterschwarz — APPROVED** [review]

**Pull request merged**